plugins/logitech-tap/fu-logitech-tap-hdmi-device.c

```diff
@@ -0,0 +1,522 @@
+				     FuProgress *progress,
+				     GError **error)
+{
+	FuLogitechTapHdmiDevice *self = FU_LOGITECH_TAP_HDMI_DEVICE(device);
+
+	/* init */
+	if (!fu_logitech_tap_hdmi_device_ait_initiate_update(self, error))
+		return FALSE;
+
+	/* write */
+	fu_progress_set_id(progress, G_STRLOC);
+	fu_progress_set_steps(progress, chunks->len);
+	for (guint i = 0; i < chunks->len; i++) {
+		FuChunk *chk = g_ptr_array_index(chunks, i);
+		g_autoptr(GByteArray) data_pkt = g_byte_array_new();
+		/* device expect fixed size buffer, so cannot leverage actual size here:
+		 * fu_chunk_get_data_sz(chk) */
+		g_byte_array_append(data_pkt, fu_chunk_get_data(chk), kLogiDefaultImageBlockSize);
+		if (!fu_logitech_tap_hdmi_device_set_xu_control(self,
+								kLogiUnitIdVidCapExtension,
+								kLogiUvcXuAitCustomCsSetFwData,
+								data_pkt->len,
+								(guchar *)data_pkt->data,
+								error))
+			return FALSE;
+		fu_progress_step_done(progress);
+	}
+
+	/* uninit */
+	if (!fu_logitech_tap_hdmi_device_ait_finalize_update(self, error))
+		return FALSE;
+
+	/* signal for sensor device to trigger composite device reboot */
+	fu_device_add_private_flag(device, FU_LOGITECH_TAP_HDMI_DEVICE_FLAG_NEEDS_REBOOT);
+	return TRUE;
+}
+
+static gboolean
+fu_logitech_tap_hdmi_device_write_firmware(FuDevice *device,
+					   FuFirmware *firmware,
+					   FuProgress *progress,
+					   FwupdInstallFlags flags,
+					   GError **error)
+{
+	g_autofree gchar *old_firmware_version = NULL;
+	g_autoptr(GBytes) fw = NULL;
+	g_autoptr(GPtrArray) chunks = NULL;
+
+	/* save the current firmware version for troubleshooting purpose */
+	old_firmware_version = g_strdup(fu_device_get_version(device));
+
+	g_debug("update %s firmware", old_firmware_version);
+
+	/* progress */
+	fu_progress_set_id(progress, G_STRLOC);
+	fu_progress_add_step(progress, FWUPD_STATUS_DEVICE_WRITE, 100, "write");
+
+	/* get image */
+	fw = fu_firmware_get_bytes(firmware, error);
+	if (fw == NULL)
+		return FALSE;
+
+	fu_progress_set_status(progress, FWUPD_STATUS_DEVICE_WRITE);
+	chunks = fu_chunk_array_new_from_bytes(fw, 0x0, 0x0, 32);
+
+	/* write */
+	if (!fu_logitech_tap_hdmi_device_write_fw(device,
+						  chunks,
+						  fu_progress_get_child(progress),
+						  error))
+		return FALSE;
+	fu_progress_step_done(progress);
+
+	/* success */
+	return TRUE;
+}
+
+static gboolean
+fu_logitech_tap_hdmi_device_set_version(FuDevice *device, GError **error)
+{
+	FuLogitechTapHdmiDevice *self = FU_LOGITECH_TAP_HDMI_DEVICE(device);
+	guint16 data_len = 0;
+	g_autofree guint8 *query_data = NULL;
+	g_autofree gchar *version_str = NULL;
+	guint8 set_data[XU_INPUT_DATA_LEN] = {kLogiTapHdmiVerSetData, 0, 0, 0, 0, 0, 0, 0};
+
+	g_debug("get HDMI firmware version");
+
+	if (!fu_logitech_tap_hdmi_device_set_xu_control(self,
+							kLogiUnitIdVidCapExtension,
+							kLogiTapCameraVersionSelector,
+							XU_INPUT_DATA_LEN,
+							(guchar *)set_data,
+							error))
+		return FALSE;
+
+	data_len = 0;
```

Comment:
i think coverity is going to complain here as it's already set to 0

Reply:
Removed